libdnf/transaction/Swdb.cpp

```diff
@@ -55,22 +57,52 @@ Swdb::Swdb(const std::string &path)
   , autoClose(true)
 {
     if (path == ":memory:") {
-        // writing to an in-memory database
-        conn = std::make_shared< SQLite3 >(path);
+        // connect to an in-memory database as requested
+        conn = std::make_shared<SQLite3>(path);
         Transformer::createDatabase(conn);
-    } else if (!pathExists(path.c_str())) {
-        // writing to a file that doesn't exist and must be created
-
-        // extract persistdir from path - "/var/lib/dnf/"
-        auto found = path.find_last_of("/");
-
-        Transformer transformer(path.substr(0, found), path);
-        transformer.transform();
-
-        conn = std::make_shared< SQLite3 >(path);
+    } else if (pathExists(path.c_str())) {
+        try {
+            // connect to an existing database
+            conn = std::make_shared<SQLite3>(path);
+            // try to execute a select to detect if the database is not read-only
+            conn->exec("BEGIN; UPDATE config SET value='test' WHERE key='test'; ROLLBACK;");
+        } catch (std::exception & ex) {
+            // re-throw the exception if running under root
+            if (geteuid() == 0) {
+                auto logger(libdnf::Log::getLogger());
+                logger->error(tfm::format("History database is not writable: %s. Error: %s", path, ex.what()));
+                throw;
+            }
+            // exec failed, connect to an in-memory database instead
+            conn = std::make_shared<SQLite3>(":memory:");
+            Transformer::createDatabase(conn);
```

Comment:
It transforms some data from previous DB or only creates an empty dataset?

Reply:
Creates an empty db.
Migration makes sense when you're running under root and you migrate from old system database.